File: BridgeX/app/src/main/java/cz/lriedel/bridgex/AuthenticationService.java
package cz.lriedel.bridgex;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;

import androidx.annotation.Nullable;

import com.google.gson.GsonBuilder;

import java.io.IOException;

import okhttp3.ResponseBody;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class AuthenticationService {

    private static final String AUTHENTICATION_PREFERENCES_NAME = "AuthenticationPreferences";
    private static final String REFRESH_TOKEN_KEY = "RefreshToken";

    private final SharedPreferences sharedPreferences;
    private final AuthenticationClient authenticationClient;

    public AuthenticationService(Context context) {
        this.sharedPreferences = context.getSharedPreferences(AUTHENTICATION_PREFERENCES_NAME, Context.MODE_PRIVATE);
        this.authenticationClient = new Retrofit.Builder()
                .baseUrl(cz.lriedel.bridgex.BuildConfig.IAM_BASE_URL)
                .addConverterFactory(GsonConverterFactory.create(new GsonBuilder().setLenient().create()))
                .build()
                .create(AuthenticationClient.class);
    }

    public void setRefreshToken(String refreshToken) {
        sharedPreferences.edit().putString(REFRESH_TOKEN_KEY, refreshToken).apply();
    }

    @Nullable
    public String getAccessToken() {
        // TODO: Introduce caching of access token if needed.
        String refreshToken = sharedPreferences.getString(REFRESH_TOKEN_KEY, null);
        if (refreshToken == null) {
            return null;
        }

        try {
            Response<AccessToken> accessTokenResponse = authenticationClient.createAccessToken(cz.lriedel.bridgex.BuildConfig.IAM_BASE_URL,
                    new AccessTokenRequest(refreshToken)).execute();
            if (!accessTokenResponse.isSuccessful()) {
                return null;
            }

            AccessToken accessToken = accessTokenResponse.body();
            if (accessToken == null) {
                return null;
            }

            setRefreshToken(accessToken.refreshToken());

            return accessToken.accessToken();
        } catch (IOException e) {
            return null;
        }
    }
}

File: BridgeX/app/src/main/java/cz/lriedel/bridgex/authentication/AuthenticationService.kt
package cz.lriedel.bridgex.authentication

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKeys
import cz.lriedel.bridgex.IamClient
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

class AuthenticationService private constructor(context: Context) {
    private val sharedPreferences: SharedPreferences = EncryptedSharedPreferences.create(
            AUTHENTICATION_PREFERENCES_NAME, MasterKeys.getOrCreate(MasterKeys.AES256_GCM_SPEC),
            context.applicationContext, EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM)

    private val iamClient: IamClient = IamClient.getOrCreate()

    private val tokenMutex = Mutex()

    private var cachedAccessToken: String? = null
    private var cachedAccessTokenExpiration: Long = 0L

    suspend fun login(username: String?, password: String?) {
        Log.d(AuthenticationService::class.java.simpleName, "Logging in as $username...")
        val iamResponse = iamClient.createToken(TokenRequest(username, password, null, DEFAULT_TOKEN_SCOPE))
        extractIamResponse(iamResponse)
    }

    fun logout() {
        extractIamResponse(null)
    }

    suspend fun getAccessToken(): String? {
        if (cachedAccessToken != null && System.currentTimeMillis() < cachedAccessTokenExpiration) {
            return cachedAccessToken
        }

        return tokenMutex.withLock {
            if (cachedAccessToken != null && System.currentTimeMillis() < cachedAccessTokenExpiration) {
                return@withLock cachedAccessToken
            }

            Log.d(AuthenticationService::class.java.simpleName, "Received a request to obtain an access token...")
            
            val refreshToken = sharedPreferences.getString(REFRESH_TOKEN_KEY, null) ?: return@withLock null

            return@withLock try {
                val iamResponse = iamClient.createToken(TokenRequest(null, null, refreshToken, DEFAULT_TOKEN_SCOPE))
                extractIamResponse(iamResponse)
                iamResponse.accessToken
            } catch (e: Exception) {
                Log.e(AuthenticationService::class.java.simpleName, "An error occurred when obtaining an access token.", e)
                null
            }
        }
    }

    private fun extractIamResponse(iamResponse: IamResponse?) {
        if (iamResponse == null) {
            cachedAccessToken = null
        }
        else {            
            cachedAccessToken = iamResponse.accessToken
            cachedAccessTokenExpiration = System.currentTimeMillis() + (iamResponse.expiresIn * 1000 * ACCESS_TOKEN_VALIDITY_MULTIPLIER).toLong()
            
            with(sharedPreferences.edit()) {
                iamResponse.refreshToken?.let { putString(REFRESH_TOKEN_KEY, it) } ?: remove(REFRESH_TOKEN_KEY)
                apply()
            }
        }
    }

    companion object {
        private const val AUTHENTICATION_PREFERENCES_NAME = "AuthenticationPreferences"
        private const val REFRESH_TOKEN_KEY = "RefreshToken"
        private const val DEFAULT_TOKEN_SCOPE = "openid offline_access"
        private const val ACCESS_TOKEN_VALIDITY_MULTIPLIER = 0.95

        @Volatile
        private var INSTANCE: AuthenticationService? = null

        fun getOrCreate(context: Context): AuthenticationService {
            return INSTANCE ?: synchronized(this) {
                val instance = INSTANCE ?: AuthenticationService(context.applicationContext)
                INSTANCE = instance
                instance
            }
        }
    }
}
